Postgres: enhance NUMERIC/DECIMAL parsing to support negative scale

IndexSeek opened the pull request:
Closes #1923 

> Beginning in PostgreSQL 15, it is allowed to declare a numeric column with a negative scale.

**Source:** [PostgreSQL Documentation - Numeric Types](https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL)

## Changes
- Modified `ExactNumberInfo::PrecisionAndScale` to use `i64` instead of `u64` for scale
- Added `parse_scale_value()` method to handle positive, negative, and explicitly positive scale values
- Updated tests to cover negative scale scenarios

iffyio commented on src/parser/mod.rs:
```diff
@@ -11229,6 +11229,38 @@ impl<'a> Parser<'a> {
+    /// - `NUMERIC(5, 2)` stores up to 5 digits with 2 decimal places (e.g., 123.45)
+    /// - `NUMERIC(5, -2)` stores up to 5 digits rounded to hundreds (e.g., 12300)
+    fn parse_scale_value(&mut self) -> Result<i64, ParserError> {
+        let next_token = self.next_token();
+        match next_token.token {
+            Token::Number(s, _) => Self::parse::<i64>(s, next_token.span.start),
+            Token::Minus => {
+                let next_token = self.next_token();
+                match next_token.token {
+                    Token::Number(s, _) => {
+                        let positive_value = Self::parse::<i64>(s, next_token.span.start)?;
+                        Ok(-positive_value)
+                    }
+                    _ => self.expected("number after minus", next_token),
+                }
+            }
+            Token::Plus => {
+                let next_token = self.next_token();
+                match next_token.token {
+                    Token::Number(s, _) => Self::parse::<i64>(s, next_token.span.start),
+                    _ => self.expected("number after plus", next_token),
+                }
+            }
+            _ => self.expected("number", next_token),
+        }
```
I think we might be able to simplify this to something like the following?
```rust
if !self.consume_token(Token::Minus) {
    return i64::try_from(self.parse_literal_uint()?)
}

let next_token = self.next_token_ref();
match &next_token.token {
  Token::Number(s, _) => Self::parse::<i64>(s, next_token.span.start),
  _ => self.expected_ref("literal int", next_token),
}
```

IndexSeek replied:
Wow, yes, this is much simpler and a way better way to handle the signs. 

Thank you!

iffyio commented on src/parser/mod.rs:
```diff
@@ -11229,6 +11229,38 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Parse a scale value for NUMERIC/DECIMAL data types.
+    ///
+    /// Supports positive, negative, and explicitly positive (with `+`) scale values.
+    /// Negative scale values are particularly useful for PostgreSQL, where they indicate
+    /// rounding to the left of the decimal point. For example:
+    /// - `NUMERIC(5, 2)` stores up to 5 digits with 2 decimal places (e.g., 123.45)
+    /// - `NUMERIC(5, -2)` stores up to 5 digits rounded to hundreds (e.g., 12300)
+    fn parse_scale_value(&mut self) -> Result<i64, ParserError> {
```
```suggestion
    /// Parse an optionally signed integer literal.
    fn parse_signed_integer(&mut self) -> Result<i64, ParserError> {
```
Think we can make the description and function name specific since it should be a straight-forward use case.

iffyio commented on src/parser/mod.rs:
```diff
@@ -17319,6 +17351,75 @@ mod tests {
+
+        #[test]
+        fn test_numeric_negative_scale() {
+            let dialect = TestedDialects::new(vec![
```
I think the negative test cases can be merged into `test_ansii_exact_numeric_types`? since they're the same feature. Also would let us avoid the duplicate tests between the two functions

IndexSeek replied:
I agree, this is a bit redundant.

IndexSeek commented:
Thank you so much for the detailed review @iffyio! I think I have addressed your comments. Please let me know if there are any other changes I should make.

IndexSeek commented on src/parser/mod.rs:
```diff
@@ -11229,6 +11229,30 @@ impl<'a> Parser<'a> {
+                (1, number_token)
+            }
+            _ => (1, next_token),
+        };
```
Thanks for the suggestion! It does read cleaner not starting with negation, looking back, opening with `if !self.consume_token(&Token::Minus)` felt a bit strange. 

I like how your solution still handles the + sign, but I noticed a few issues with the logic flow. I think that with an optional plus token `let _ = self.consume_token(&Token::Plus);`, it calls self.advance_token() and then self.get_current_token(), which could advance twice and potentially skip the actual number token. 

I did take a look at `peek_token_ref` and was wondering if we could use it like so?

```rs
    fn parse_signed_integer(&mut self) -> Result<i64, ParserError> {
        let is_negative = self.consume_token(&Token::Minus);

        if !is_negative {
            let _ = self.consume_token(&Token::Plus);
        }

        let current_token = self.peek_token_ref();
        match &current_token.token {
            Token::Number(s, _) => {
                let s = s.clone();
                let span_start = current_token.span.start;
                self.advance_token();
                let value = Self::parse::<i64>(s, span_start)?;
                Ok(if is_negative { -value } else { value })
            }
            _ => self.expected_ref("number", current_token),
        }
    }
```

I think that should avoid us needing to consume the token until we know we have a number.

IndexSeek replied:
Should we still parse the "+" character or error out? I know I originally added a test for it but I'm not sure any dialects support a "+" token in the scale definition for numeric types.

IndexSeek replied:
That sounds good to me! I also added back that test that was explicitly checking to ensure that would work and be removed, but we can remove those lines if it is redundant.

iffyio reviewed (approved):
LGTM! Thanks @IndexSeek!
cc @alamb
